vm/src/vm/runners/cairo_pie.rs

```diff
@@ -85,6 +86,45 @@ pub enum BuiltinAdditionalData {
     None,
 }
 
+impl PartialEq for BuiltinAdditionalData {
+    fn eq(&self, other: &BuiltinAdditionalData) -> bool {
+        match self {
+            BuiltinAdditionalData::Empty(_) => match other {
+                BuiltinAdditionalData::Empty(_) => true,
+                BuiltinAdditionalData::Hash(hashes) => hashes.is_empty(),
+                BuiltinAdditionalData::Signature(signatures) => signatures.is_empty(),
+                _ => false,
+            },
+            BuiltinAdditionalData::Hash(data) => {
+                if let BuiltinAdditionalData::Hash(other_data) = other {
+                    data == other_data
+                } else if let BuiltinAdditionalData::Empty(_) = other {
+                    data.is_empty()
+                } else {
+                    false
+                }
+            }
+            BuiltinAdditionalData::Signature(data) => {
+                if let BuiltinAdditionalData::Signature(other_data) = other {
+                    data == other_data
+                } else if let BuiltinAdditionalData::Empty(_) = other {
+                    data.is_empty()
+                } else {
+                    false
+                }
+            }
+            BuiltinAdditionalData::Output(data) => {
+                if let BuiltinAdditionalData::Output(other_data) = other {
+                    data == other_data
+                } else {
+                    false
+                }
+            }
+            BuiltinAdditionalData::None => matches!(other, BuiltinAdditionalData::None),
+        }
+    }
+}
```

Comment:
I think this could be simplified by:
1. Creating an `is_empty` method that checks only that for all variants.
2. Then doing that check first (with early return `true` if both are empty and early return `false` if one is but the other isn't) and otherwise requires both the variant and data to match, which can be done with a single `match` for both arguments.

Reply:
E.g.:
```suggestion
impl BuiltinAdditionalData {
    fn is_empty(&self) -> bool {
        use BuiltinAdditionalData::*;
        match self {
            Empty => true,
            Hash(data) => data.is_empty(),
            Signature(data) => data.is_empty(),
            Output(data) => data.is_empty(),
            None => true,
        }
    }
}

impl PartialEq for BuiltinAdditionalData {
    fn eq(&self, other: &BuiltinAdditionalData) -> bool {
        use BuiltinAdditionalData::*;
        if matches!(self, BuiltinAdditionalData::None) {
            return matches!(other, BuiltinAdditionalData::None);
        }
        match (self.is_empty(), other.is_empty()) {
            (true, true) => { return true; },
            (true, false)|(false, true) => { return false; },
            _ => (),
        }
        match (self, other) {
            (Hash(data), Hash(other_data)) if data == other_data => true,
            (Signature(data), Signature(other_data)) if data == other_data => true,
            (Output(data), Output(other_data)) if data == other_data => true,
            _ => false,
        }
    }
}
```

Reply:
This doesn't distinguish an empty hash list from an empty signature list, but I agree that that's a good sacrifice for clarity.

Reply:
Huh, I misread the original code as (intentionally) doing that. Now that I read it again I see you use if let to distinguish.